pkg/metrics/metrics.go

```diff
@@ -0,0 +1,174 @@
+package metrics
+
+import (
+	"encoding/json"
+	"strings"
+
+	corelisters "k8s.io/client-go/listers/core/v1"
+
+	"github.com/prometheus/client_golang/prometheus"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+
+	configv1 "github.com/openshift/cluster-samples-operator/pkg/apis/samples/v1"
+	"github.com/openshift/cluster-samples-operator/pkg/client"
+	sampoplisters "github.com/openshift/cluster-samples-operator/pkg/generated/listers/samples/v1"
+
+	"github.com/sirupsen/logrus"
+)
+
+const (
+	// per https://prometheus.io/docs/practices/naming/#metric-names feel our metrics here fall under the _info category of classification
+
+	// since we want to index by imagestream name, where each imagestream is either failing or not failing, this is why
+	// we went with the _info suffix vs. the _total suffix
+	failedImportsQuery = "openshift_samples_failed_imagestream_import_info"
+
+	degradedQuery = "openshift_samples_degraded_info"
+
+	invalidConfigQuery = "openshift_samples_invalidconfig_info"
+
+	invalidSecretQuery = "openshift_samples_invalidsecret_info"
+
+	missingSecret        = "missing_secret"
+	missingTBRCredential = "missing_tbr_credential"
+
+	// MetricsPort is the IP port supplied to the HTTP server used for prometheus,
+	// and matches what is specified in the corresponding Service and ServiceMonitor
+	MetricsPort = 60000
+)
+
+var (
+	importsFailedDesc = prometheus.NewDesc(failedImportsQuery,
+		"Indicates by name whether a samples imagestream import has failed or not (1 == failure, 0 == succes).",
+		[]string{"name"},
+		nil)
+	invalidSecretDesc = prometheus.NewDesc(invalidSecretQuery,
+		"Indicates if the pull secret is functional in the openshift namespace.  If not functional, the reason supplied either points to it missing in the openshift namesapce, or if it is missing credentials for registry.redhat.io when they should exist (1 == not functional, 0 == functional).",
+		[]string{"reason"},
+		nil)
+	degradedStat = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: degradedQuery,
+		Help: "Indicates if the samples operator is degraded.",
+	})
+	configInvalidStat = prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: invalidConfigQuery,
+		Help: "Indicates the configuration for the samples operator has an invalid configuration setting.",
+	})
+
+	sc         = samplesCollector{}
+	registered = false
+
+	// Streams is the list of streams from the install payload that is maintained by the Handler
+	Streams = []string{}
```

Comment:
Why a package-global variable?

Reply:
it is populated in files.go and referenced in handler.go

Reply:
I've changed it to `streams` and added some public helper methods